### PR TITLE
Add unit tests and improve code coverage.

### LINKED ISSUE #214: Several ocean.py modules don't have have test files
TODO for this issue: make sure that there _is_ a test file for each module.

Note: this is a sub-issue of the epic in #149. The epic gives a more thorough description / context.

### LINKED ISSUE #256: test_generate_multi_value_hash failure
## Failure Information

`ocean_lib/web3_internal/test/test_utils.py::test_generate_multi_value_hash` fails when using the `TEST_PRIVATE_KEY1` value listed in `marketplace-flow.md` and `developers.md`.  The test passes when using the `TEST_PRIVATE_KEY1` value listed in `pytest.ini` or `tox.ini`.

## Expected Behavior

All tests pass.

## Step to Reproduce the problem

1. Start barge, export TEST_PRIVATE_KEY1 env variable, deploy fake OCEAN as described in the [developers flow](https://github.com/oceanprotocol/ocean.py/blob/main/READMEs/developers.md)
2. Execute `pytest ocean_lib/web3_internal/test/test_utils.py::test_generate_multi_value_hash`

### Possible Solution

Change `marketplace-flow.md` and `developers.md` such that `TEST_PRIVATE_KEY1` and `TEST_PRIVATE_KEY2` matches the values listed in `pytest.ini` and `tox.ini`.

## Reference

`marketplace-flow.md` and `developers.md`
```
export TEST_PRIVATE_KEY1=0xc594c6e5def4bab63ac29eed19a134c130388f74f019bc74b8f4389df2837a58
export TEST_PRIVATE_KEY2=0xef4b441145c1d0f3b4bc6d61d29f5c6e502359481152f869247c7a4244d45209
```

`pytest.ini` and `tox.ini`
```
    TEST_PRIVATE_KEY1=0xbbfbee4961061d506ffbb11dfea64eba16355cbf1d9c29613126ba7fec0aed5d
    TEST_PRIVATE_KEY2=0x804365e293b9fab9bd11bddd39082396d56d30779efbb3ffb0a6089027902c4a
```

### LINKED ISSUE #273: failing test: test_generate_multi_value_hash()
**Describe the bug**

Running
```console
pytest ocean_lib/web3_internal/test/test_utils.py::test_generate_multi_value_hash
```

Will cause
```console
AssertionError: assert '0x7ba270cc76...1b60662f60cea' == '0x6d59f15c58...00c0d736a9fec'
```

**To Reproduce**
1. Go through steps in developers.md
2. When running all tests with `pytest`, it fails on the above.

To reproduce faster, it's the test given above.

**Expected behavior**
Don't fail.

**Logs**

```console
(venv) trentmc@trentmcLnv:~/code/ocean.py$ pytest ocean_lib/web3_internal/test/test_utils.py
================================================================= test session starts ==================================================================
platform linux -- Python 3.8.8, pytest-6.2.3, py-1.10.0, pluggy-0.13.1
rootdir: /home/trentmc/code/ocean.py, configfile: pytest.ini
plugins: env-0.6.2
collected 2 items                                                                                                                                      

ocean_lib/web3_internal/test/test_utils.py F.                                                                                                    [100%]

======================================================================= FAILURES =======================================================================
____________________________________________________________ test_generate_multi_value_hash ____________________________________________________________

alice_address = '0xe2DD09d719Da89e5a3D0F2549c7E24566e947260'

    def test_generate_multi_value_hash(alice_address):
        with pytest.raises(AssertionError):
            generate_multi_value_hash(["more", "types", "than"], ["values"])
    
        expected = "0x6d59f15c5814d9fddd2e69d1f6f61edd0718e337c41ec74011900c0d736a9fec"
>       assert generate_multi_value_hash(["address"], [alice_address]).hex() == expected
E       AssertionError: assert '0x7ba270cc76...1b60662f60cea' == '0x6d59f15c58...00c0d736a9fec'
E         - 0x6d59f15c5814d9fddd2e69d1f6f61edd0718e337c41ec74011900c0d736a9fec
E         + 0x7ba270cc76c2dde25e744613ec459be48c8130f6f996b66f8df1b60662f60cea

ocean_lib/web3_internal/test/test_utils.py:17: AssertionError
---------------------------------------------------------------- Captured stdout setup -----------------------------------------------------------------
sender: 0xc594c6e5def4bab63ac29eed19a134c130388f74f019bc74b8f4389df2837a58, 0xe2DD09d719Da89e5a3D0F2549c7E24566e947260, None, address: 0xe2DD09d719Da89e5a3D0F2549c7E24566e947260
private key: 0xc594c6e5def4bab63ac29eed19a134c130388f74f019bc74b8f4389df2837a58
public key: 0x8b1b222564fb8302e9acfa4bcd60060de03ae479fcbed38fe21e45cd39ddb3fefb60ba1b4625e00d02223027c8f485f6d77693ab62ce06289b23942c6b319ccf

sender balance: 98.696703797
=============================================================== short test summary info ================================================================
FAILED ocean_lib/web3_internal/test/test_utils.py::test_generate_multi_value_hash - AssertionError: assert '0x7ba270cc76...1b60662f60cea' == '0x6d59f...
============================================================= 1 failed, 1 passed in 1.47s ==============================================================
```

### PR DESCRIPTION
**Pull request by mariacarmina**
Towards #214
Closes #273
Closes #256

Changes proposed in this PR:

- add unit tests for the unused/untested functions;
- code coverage above 90% for each file;
- resolve `generate_hash` bug in the web3 unit tests.

### CODE REVIEW

File: ocean_lib/assets/test/test_asset.py
Comment (calina-c) [inline ocean_lib/assets/test/test_asset.py:27]: you can also test the format, e.g. length, starts with 0x (if that is the case). It would be a bit more thorough.

File: ocean_lib/assets/test/test_asset.py
Comment (calina-c) [inline ocean_lib/assets/test/test_asset.py:53]: please check the contents as well (checksum contents). Otherwise, great job!

**Review by calina-c — CHANGES_REQUESTED** [review]

File: ocean_lib/assets/test/test_asset.py
Comment (mariacarmina) [reply ocean_lib/assets/test/test_asset.py:53]: Sure, thanks!

File: ocean_lib/web3_internal/test/test_utils.py
Comment (trentmc) [inline ocean_lib/web3_internal/test/test_utils.py:12]: I just tested this, in this branch, and it works now. +1

**Comment by trentmc** [comment]
I just tested to see that it solves #273 in this branch. And, it does. Good job:)

I briefly reviewed the other code. Looks good, but I defer to Calina and David for a more thorough review.

File: ocean_lib/assets/test/test_asset.py
Comment (calina-c) [inline ocean_lib/assets/test/test_asset.py:56]: you can use enumerate instead of range, for simplicity.

File: ocean_lib/assets/test/test_utils.py
Comment (calina-c) [inline ocean_lib/assets/test/test_utils.py:19]: can you please add docstrings to all your tests?

File: ocean_lib/assets/test/test_utils.py
Comment (calina-c) [inline ocean_lib/assets/test/test_utils.py:60]: no need for the assignment, if it's not used. Just remove the "_ =" part.

File: ocean_lib/assets/test/test_asset.py
Comment (calina-c) [inline ocean_lib/assets/test/test_asset.py:28]: you can drop the "is True" part. `assert` will do just that.

**Review by calina-c — CHANGES_REQUESTED** [review]

File: ocean_lib/assets/test/test_asset.py
Comment (mariacarmina) [reply ocean_lib/assets/test/test_asset.py:56]: But it is needed for Codacy too, will do it.

File: ocean_lib/assets/test/test_utils.py
Comment (calina-c) [inline ocean_lib/assets/test/test_utils.py:49]: should be "Tests..." something, handling is not for tests, but codebase itself.

File: ocean_lib/assets/test/test_asset.py
Comment (calina-c) [inline ocean_lib/assets/test/test_asset.py:17]: can you be a bit more specific here?

File: tox.ini
Comment (calina-c) [inline tox.ini:14]: Don't change private keys, please! We rely on these in various places.

**Review by calina-c — CHANGES_REQUESTED** [review]
2 nitpicks and a problem. Otherwise looking good.

File: tox.ini
Comment (mariacarmina) [reply tox.ini:14]: I talked with @trentmc to change those keys regarding to the web3 unit test.

File: ocean_lib/web3_internal/test/test_utils.py
Comment (calina-c) [reply ocean_lib/web3_internal/test/test_utils.py:12]: Agreed, and that's what I suggested as well.

Comment (calina-c) [reply ocean_lib/web3_internal/test/test_utils.py:12]: And @mariacarmina to make the test pass in all instances, just get the value from env instead of hardcoding it. E.g. expected = os.getenv('TEST_PRIVATE_KEY1') instead of "0x6..."

Comment (mariacarmina) [reply ocean_lib/web3_internal/test/test_utils.py:12]: Agreed, thank you @DMats and @calina-c.

Comment (mariacarmina) [reply ocean_lib/web3_internal/test/test_utils.py:12]: Modified the test keys in pytest.ini to match tox.ini.

Comment (mariacarmina) [reply ocean_lib/web3_internal/test/test_utils.py:12]: Also, @DMats in your code snippet, the assert related to `alice_private_key` will fail, because now it's another test key for that (the one from the comment).

File: ocean_lib/ocean/test/test_ocean_assets.py
Comment (DMats) [inline ocean_lib/ocean/test/test_ocean_assets.py:248]: This assertion only checks whether the asset was created. My gut reaction is that more assertions are warranted. But after digging into `OceanAssets.create()` I realized that the `owner_address` parameter is silently ignored if `data_token_address` parameter is set, so this test is instruction-for-instruction identical to `test_create_asset_with_address`. In the end, I concluded that I don't know how to improve this test, but I wanted to write this comment anyway because I spent so much time on it.

File: ocean_lib/ocean/test/test_ocean_assets.py
Comment (DMats) [inline ocean_lib/ocean/test/test_ocean_assets.py:268]: The purpose of the `owner_address` parameter is to specify an address that is different from the publisher address so that ownership can be transferred immediately after the asset is published. 

Recommended actions:

1. Duplicate this test.
2. Change the first test so that the `owner_address` parameter is different from the publisher address.
3. Keep the second test the same.

Reasoning:
This tests both the intended and unintended usage of the method.

File: ocean_lib/ocean/test/test_ocean_assets.py
Comment (DMats) [inline ocean_lib/ocean/test/test_ocean_assets.py:277]: Recommend renaming to be more expressive like "test_pay_for_service_insufficient_balance"

**Review by DMats — CHANGES_REQUESTED** [review]

File: ocean_lib/assets/test/test_asset.py
Comment (calina-c) [reply ocean_lib/assets/test/test_asset.py:19]: typically, yes. But in practice using metadata directly makes tests fail. Making a copy is the easiest solution for testing right now.

File: ocean_lib/assets/test/test_utils.py
Comment (calina-c) [reply ocean_lib/assets/test/test_utils.py:1]: I don't think we gave much thought to that. Since these functions are helpers to deal with medium/big JSONs, I don't see removing an algo that doesn't exist as a security risk or blocker. It depends on how we want to deal with it:
- block the flow and spit out an error?
- fail silently?

Either way I'd make a separate issue for that, since this one is getting crowded (I think there's about 3 issues solved with one PR)

@trentmc what do you think?

File: ocean_lib/assets/test/test_utils.py
Comment (calina-c) [reply ocean_lib/assets/test/test_utils.py:53]: It's not quite the same thing. From memory without checking 100% I think the difference is:

- wait_for_ddo waits until it's published
- resolve assumes it's been published and only tries to resolve (decrypt it) -> sometimes used to double check whether the publishing went well and gives no error

**Comment by trentmc** [comment]
I chatted with Calina. Result:

It's to change the private keys in the READMEs.

However do be sure that you can run through the quickstart.

File: ocean_lib/assets/test/test_asset.py
Comment (DMats) [reply ocean_lib/assets/test/test_asset.py:19]: Okay. I'll ignore this for now.

**Comment by alexcos20** [comment]
closes #256 as well

File: ocean_lib/assets/test/test_utils.py
Comment (mariacarmina) [reply ocean_lib/assets/test/test_utils.py:53]: `wait_for_ddo` tries to do the `resolve` part after the asset is published. Will change to `wait_for_ddo` for consistency.

File: ocean_lib/ocean/test/test_ocean_assets.py
Comment (mariacarmina) [reply ocean_lib/ocean/test/test_ocean_assets.py:248]: So, @DMats do you think it's useful to keep it or do you want to include it in `test_create_asset_with_address, because it's kind of similar? The difference consists in setting the parameters.

Comment (DMats) [reply ocean_lib/ocean/test/test_ocean_assets.py:248]: If it were up to me, I'd remove the `owner_address` parameter from `OceanAssets.create()`. Then I'd remove this test (`test_create_asset_with_owner_address`). That said, I'd wait for confirmation from @calina-c or @trentmc to proceed.

Reasoning:

1. As far as I can tell, `owner_address` is not used in ocean.py or anywhere else in the ocean software stack. 
2. The [corresponding method in ocean.js](https://github.com/oceanprotocol/ocean.js/blob/main/src/ocean/Assets.ts#L69) does not have the `owner_address` parameter.

Caveats:

This would be an API-breaking change (assuming that OceanAssets is considered part of ocean.py's API).

Comment (mariacarmina) [reply ocean_lib/ocean/test/test_ocean_assets.py:248]: This will cause a breaking change of the interface if I will remove the `owner_address`. Been there, done that before.

Comment (calina-c) [reply ocean_lib/ocean/test/test_ocean_assets.py:244]: keep this, it was preciously discussed. `owner_address` stays :D

File: ocean_lib/assets/test/test_utils.py
Comment (calina-c) [reply ocean_lib/assets/test/test_utils.py:1]: @trentmc Some guidance, please. We forgot about this when we discussed this PR.

File: ocean_lib/web3_internal/test/test_utils.py
Comment (calina-c) [reply ocean_lib/web3_internal/test/test_utils.py:14]: @mariacarmina can you confirm you ran the quickstart and everything is ok?

File: ocean_lib/ocean/test/test_ocean_assets.py
Comment (calina-c) [reply ocean_lib/ocean/test/test_ocean_assets.py:263]: @DMats that is covered in integration tests, but if @mariacarmina has the time, I'd suggest we cover it here, too.

File: ocean_lib/ocean/test/test_ocean_assets.py
Comment (DMats) [reply ocean_lib/ocean/test/test_ocean_assets.py:244]: will be addressed in #284

File: ocean_lib/web3_internal/test/test_utils.py
Comment (mariacarmina) [reply ocean_lib/web3_internal/test/test_utils.py:14]: Done.

File: ocean_lib/assets/test/test_asset.py
Comment (DMats) [inline ocean_lib/assets/test/test_asset.py:53]: I think returning `None` from the test will cause the rest of the assert statements to be skipped. Please replace this if with `assert privacy_dict`

**Review by DMats — CHANGES_REQUESTED** [review]

File: ocean_lib/assets/test/test_asset.py
Comment (calina-c) [reply ocean_lib/assets/test/test_asset.py:53]: I agree. I missed this, but at this point there should be a privacy dict there. Yes, please follow David's suggestion.

File: ocean_lib/assets/test/test_utils.py
Comment (trentmc) [reply ocean_lib/assets/test/test_utils.py:1]: Separate issue is best, yes.

File: ocean_lib/assets/test/test_asset.py
Comment (mariacarmina) [reply ocean_lib/assets/test/test_asset.py:53]: Done

File: ocean_lib/ocean/test/test_ocean_assets.py
Comment (DMats) [reply ocean_lib/ocean/test/test_ocean_assets.py:263]: will be addressed in #284

File: ocean_lib/assets/test/test_utils.py
Comment (DMats) [reply ocean_lib/assets/test/test_utils.py:1]: will be addressed in https://github.com/oceanprotocol/ocean.py/issues/284

File: ocean_lib/assets/test/test_asset.py
Comment (DMats) [reply ocean_lib/assets/test/test_asset.py:53]: @mariacarmina I'm afraid you misunderstood me.  

Please make it look like the following.  Remove the `if` and `return` statements.

```python
privacy_dict = service.attributes["main"].get("privacy")
assert privacy_dict
```

**Review by DMats — CHANGES_REQUESTED** [review]

**Review by DMats — APPROVED** [review]

**Review by calina-c — APPROVED** [review]

**Pull request merged**
